v2/navigation.jsx

```diff
@@ -276,13 +275,44 @@ export function Navigation(props) {
     }, [state.state, state.stack, state.commit])
 
     const api = {
-        push: (name, params = {}) => dispatch({ type: 'push', name, opts: { animate: props.animate }, params }),
-        pop: () => dispatch({ type: 'pop', opts: { animate: props.animate } }),
-        goto: (stack) => dispatch({ type: 'goto', stack, opts: { animate: props.animate } }),
+        /**
+         * @param {ScreenName} name
+         * @param {Record<string, any>} params
+         */
+        push: (name, params = {}) => {
+            const url = new URL(window.location.href)
+
+            for (let [key, value] of Object.entries(params)) {
+                // using 'set' to override any previous values.
+                url.searchParams.set(key, value)
+            }
+
+            // reset the navigation stack
+            url.searchParams.delete('stack')
+            for (let string of state.stack) {
+                url.searchParams.append('stack', string)
+            }
+            url.searchParams.append('stack', name)
+
+            // reflect the new as a push
+            window.history.pushState({}, '', url)
+
+            // change component state
+            dispatch({ type: 'push', name, opts: { animate: props.animate } })
```

Comment:
This is the main difference that fixes everything. iOS will take a 'snapshot' of the current UI, so by pushing to history *first*, it ensures it gets the correct screen